File: src/strategy/movingaveragestrategy.cpp
#include "movingaveragestrategy.h"
#include "../indicator/indicators.h"
#include <QDateTime>
#include <algorithm>

MovingAverageStrategy::MovingAverageStrategy(QObject *parent)
    : StrategyBase(parent)
    , m_fastMA(0.0)
    , m_slowMA(0.0)
    , m_rsi(50.0)
    , m_macd(0.0)
    , m_bollingerUpper(0.0)
    , m_bollingerLower(0.0)
{
}

void MovingAverageStrategy::reset()
{
    StrategyBase::reset();
    m_priceHistory.clear();
    m_closePrices.clear();
    m_highPrices.clear();
    m_lowPrices.clear();
    m_volumes.clear();
    m_fastMA = 0.0;
    m_slowMA = 0.0;
}

void MovingAverageStrategy::processMarketData(const MarketData &data)
{
    if (data.symbol != m_symbol) return;

    m_priceHistory.push_back(data);
    m_closePrices.push_back(data.close);
    m_highPrices.push_back(data.high);
    m_lowPrices.push_back(data.low);
    m_volumes.push_back(data.volume);

    const int maxHistory = 100;
    if (m_priceHistory.size() > maxHistory) {
        m_priceHistory.pop_front();
        m_closePrices.erase(m_closePrices.begin());
        m_highPrices.erase(m_highPrices.begin());
        m_lowPrices.erase(m_lowPrices.begin());
        m_volumes.erase(m_volumes.begin());
    }

    calculateIndicators();

    if (m_hasPosition) {
        updateStopLossTakeProfit(data.close);
    }

    checkTradingSignal(data.close);
}

StrategyRuntimeSnapshot MovingAverageStrategy::runtimeSnapshot() const
{
    StrategyRuntimeSnapshot snapshot;
    snapshot.sampleCount = static_cast<int>(m_closePrices.size());
    snapshot.requiredSamples = std::max(m_params.fastMA, m_params.slowMA);
    snapshot.readyForSignal = snapshot.requiredSamples > 0 && snapshot.sampleCount >= snapshot.requiredSamples;
    snapshot.hasFastSlowMA = snapshot.readyForSignal && m_fastMA > 0.0 && m_slowMA > 0.0;
    snapshot.fastMA = m_fastMA;
    snapshot.slowMA = m_slowMA;
    return snapshot;
}

void MovingAverageStrategy::calculateIndicators()
{
    if (m_closePrices.size() < m_params.slowMA) {
        m_fastMA = 0.0;
        m_slowMA = 0.0;
        return;
    }

    m_fastMA = TechnicalIndicators::calculateSMA(m_closePrices, m_params.fastMA);
    m_slowMA = TechnicalIndicators::calculateSMA(m_closePrices, m_params.slowMA);
    m_rsi = TechnicalIndicators::calculateRSI(m_closePrices, 14);
    auto [macd, signal] = TechnicalIndicators::calculateMACD(m_closePrices, 12, 26, 9);
    m_macd = macd;
    auto [lower, upper] = TechnicalIndicators::calculateBollingerBands(m_closePrices, 20, 2.0);
    m_bollingerLower = lower;
    m_bollingerUpper = upper;
}

void MovingAverageStrategy::checkTradingSignal(double price)
{
    if (m_closePrices.size() < m_params.slowMA) return;

    double prevFastMA = TechnicalIndicators::calculateSMA(
        std::vector<double>(m_closePrices.begin(), m_closePrices.end() - 1), m_params.fastMA);
    double prevSlowMA = TechnicalIndicators::calculateSMA(
        std::vector<double>(m_closePrices.begin(), m_closePrices.end() - 1), m_params.slowMA);

    bool goldenCross = (prevFastMA <= prevSlowMA) && (m_fastMA > m_slowMA);
    bool deathCross = (prevFastMA >= prevSlowMA) && (m_fastMA < m_slowMA);

    if (!m_hasPosition) {
        if (goldenCross && m_rsi < 70) {
            m_hasPosition = true;
            m_entryPrice = price;
            m_entryTime = QDateTime::currentDateTime();

            m_stopLossPrice = m_entryPrice * (1 - m_params.stopLossPercent / 100.0);
            m_takeProfitPrice = m_entryPrice * (1 + m_params.takeProfitPercent / 100.0);

            emitSignal(SignalType::BUY, price, m_params.lotSize,
                       QString("金叉信号: 快速MA(%1)上穿慢速MA(%2)").arg(m_params.fastMA).arg(m_params.slowMA));
        }
    } else {
        if (deathCross) {
            m_hasPosition = false;
            emitSignal(SignalType::SELL, price, m_params.lotSize,
                       QString("死叉信号: 快速MA(%1)下穿慢速MA(%2)").arg(m_params.fastMA).arg(m_params.slowMA));
        } else if (price <= m_stopLossPrice) {
            m_hasPosition = false;
            emitSignal(SignalType::STOP_LOSS, price, m_params.lotSize,
                       QString("止损触发: 亏损%1%").arg(m_params.stopLossPercent));
        } else if (price >= m_takeProfitPrice) {
            m_hasPosition = false;
            emitSignal(SignalType::TAKE_PROFIT, price, m_params.lotSize,
                       QString("止盈触发: 盈利%1%").arg(m_params.takeProfitPercent));
        }
    }
}

void MovingAverageStrategy::updateStopLossTakeProfit(double price)
{
    double currentProfit = (price - m_entryPrice) / m_entryPrice * 100.0;

    if (currentProfit >= 3.0) {
        double newSL = m_entryPrice * (1 + 0.01);
        if (newSL > m_stopLossPrice) {
            m_stopLossPrice = newSL;
        }
    }

    if (currentProfit >= 5.0) {
        double newTP = m_entryPrice * (1 + 0.10);
        if (newTP > m_takeProfitPrice) {
            m_takeProfitPrice = newTP;
        }
    }
}

File: quant/app/src/main/java/com/example/starquant/domain/strategy/MovingAverageStrategy.kt
package com.example.starquant.domain.strategy

import com.example.starquant.data.model.MarketData
import com.example.starquant.data.model.SignalType
import com.example.starquant.domain.indicator.TechnicalIndicators
import java.util.Calendar
import java.util.Date
import kotlin.math.max
import kotlin.math.min

class MovingAverageStrategy : StrategyBase() {
    private var hasCurrentBar: Boolean = false
    private var currentBarStart: Date = Date()
    private var currentBar: MarketData = MarketData()
    private val priceHistory = ArrayDeque<MarketData>()
    private val closePrices = mutableListOf<Double>()
    private val highPrices = mutableListOf<Double>()
    private val lowPrices = mutableListOf<Double>()
    private val volumes = mutableListOf<Double>()

    private var fastMA: Double = 0.0
    private var slowMA: Double = 0.0
    private var rsi: Double = 50.0
    private var macd: Double = 0.0
    private var bollingerUpper: Double = 0.0
    private var bollingerLower: Double = 0.0
    private var buyConfirmBars: Int = 0
    private var sellConfirmBars: Int = 0
    private var cooldownBarsRemaining: Int = 0

    override fun reset() {
        super.reset()
        hasCurrentBar = false
        currentBarStart = Date()
        currentBar = MarketData()
        priceHistory.clear()
        closePrices.clear()
        highPrices.clear()
        lowPrices.clear()
        volumes.clear()
        fastMA = 0.0
        slowMA = 0.0
        rsi = 50.0
        macd = 0.0
        bollingerUpper = 0.0
        bollingerLower = 0.0
        buyConfirmBars = 0
        sellConfirmBars = 0
        cooldownBarsRemaining = 0
    }

    fun closedBarCount(): Int = closePrices.size

    fun getFastMA(): Double = fastMA
    fun getSlowMA(): Double = slowMA
    fun getRSI(): Double = rsi
    fun getMACD(): Double = macd
    fun getBollingerUpper(): Double = bollingerUpper
    fun getBollingerLower(): Double = bollingerLower
    fun getBuyConfirmBars(): Int = buyConfirmBars
    fun getSellConfirmBars(): Int = sellConfirmBars
    fun getCooldownBarsRemaining(): Int = cooldownBarsRemaining
    fun hasPositionState(): Boolean = hasPosition
    fun getEntryPrice(): Double = positionEntryPrice
    fun getStopLossPrice(): Double = positionStopLossPrice
    fun getTakeProfitPrice(): Double = positionTakeProfitPrice

    override fun processMarketData(data: MarketData) {
        if (data.symbol != symbol || data.close <= 0.0) return

        val timestamp = if (data.timestamp.time > 0) data.timestamp else Date()
        val barStart = barStartFor(timestamp)

        if (!hasCurrentBar) {
            startBar(data, barStart)
            return
        }

        if (barStart <= currentBarStart) {
            updateCurrentBar(data)
            return
        }

        processClosedBar(currentBar)
        startBar(data, barStart)
    }

    private fun barPeriodMinutes(): Int {
        val minutes = strategyConfig.doubleMAConfig.barPeriodMinutes
        return if (minutes == 1 || minutes == 3 || minutes == 5) minutes else 5
    }

    private fun barStartFor(timestamp: Date): Date {
        val cal = Calendar.getInstance()
        cal.time = timestamp
        val period = barPeriodMinutes()
        val minute = (cal.get(Calendar.MINUTE) / period) * period
        cal.set(Calendar.MINUTE, minute)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    private fun startBar(data: MarketData, barStart: Date) {
        hasCurrentBar = true
        currentBarStart = barStart
        currentBar = data.copy()
        currentBar.timestamp = barStart
        currentBar.open = data.close
        currentBar.high = data.close
        currentBar.low = data.close
        currentBar.close = data.close
    }

    private fun updateCurrentBar(data: MarketData) {
        if (!hasCurrentBar || data.close <= 0.0) return
        currentBar.high = max(currentBar.high, data.close)
        currentBar.low = min(currentBar.low, data.close)
        currentBar.close = data.close
        currentBar.volume = data.volume
        currentBar.turnover = data.turnover
        if (data.averagePrice > 0.0) {
            currentBar.averagePrice = data.averagePrice
        }
        currentBar.tradeCount = data.tradeCount
    }

    private fun processClosedBar(bar: MarketData) {
        priceHistory.addLast(bar)
        closePrices.add(bar.close)
        highPrices.add(bar.high)
        lowPrices.add(bar.low)
        volumes.add(bar.volume)

        val maxHistory = 100
        if (priceHistory.size > maxHistory) {
            priceHistory.removeFirst()
            closePrices.removeAt(0)
            highPrices.removeAt(0)
            lowPrices.removeAt(0)
            volumes.removeAt(0)
        }

        calculateIndicators()

        if (hasPosition) {
            updateStopLossTakeProfit(bar.close)
        }

        if (cooldownBarsRemaining > 0) {
            cooldownBarsRemaining--
        }

        checkTradingSignal(bar)
    }

    private fun calculateIndicators() {
        if (closePrices.size < params.slowMA) {
            fastMA = 0.0
            slowMA = 0.0
            return
        }
        fastMA = TechnicalIndicators.calculateSMA(closePrices, params.fastMA)
        slowMA = TechnicalIndicators.calculateSMA(closePrices, params.slowMA)
        rsi = TechnicalIndicators.calculateRSI(closePrices, 14)
        val macdResult = TechnicalIndicators.calculateMACD(closePrices, 12, 26, 9)
        macd = macdResult.macd
        val bbResult = TechnicalIndicators.calculateBollingerBands(closePrices, 20, 2.0)
        bollingerLower = bbResult.lower
        bollingerUpper = bbResult.upper
    }

    private fun confirmationBars(): Int {
        return strategyConfig.doubleMAConfig.confirmationBars.coerceIn(1, 5)
    }

    private fun cooldownBars(): Int {
        return strategyConfig.doubleMAConfig.cooldownBars.coerceIn(0, 20)
    }

    private fun minSpreadPercent(): Double {
        return strategyConfig.doubleMAConfig.minSpreadPercent.coerceIn(0.0, 5.0)
    }

    private fun minRewardCostMultiple(): Double {
        return strategyConfig.doubleMAConfig.minRewardCostMultiple.coerceIn(0.0, 10.0)
    }

    private fun isLateBuyBlocked(barTime: Date): Boolean {
        if (!strategyConfig.doubleMAConfig.blockLateBuy) return false
        val cal = Calendar.getInstance()
        cal.time = barTime
        return cal.get(Calendar.HOUR_OF_DAY) > 14 ||
                (cal.get(Calendar.HOUR_OF_DAY) == 14 && cal.get(Calendar.MINUTE) >= 45)
    }

    private fun estimatedRoundTripCostPercent(price: Double): Double {
        if (price <= 0.0) return 0.0
        val volume = if (params.lotSize > 0.0) params.lotSize else 100.0
        val amount = price * volume
        if (amount <= 0.0) return 0.0
        val buyCommission = max(amount * 0.0002, 5.0)
        val sellCommission = max(amount * 0.0002, 5.0)
        val buyTransfer = amount * 0.00001
        val sellTransfer = amount * 0.00001
        val stampDuty = amount * 0.001
        return (buyCommission + sellCommission + buyTransfer + sellTransfer + stampDuty) / amount * 100.0
    }

    private fun enterCooldown() {
        buyConfirmBars = 0
        sellConfirmBars = 0
        cooldownBarsRemaining = cooldownBars()
    }

    private fun checkTradingSignal(bar: MarketData) {
        val price = bar.close
        if (closePrices.size <= params.slowMA || price <= 0.0) return

        val previousCloses = closePrices.dropLast(1)
        val prevSlowMA = TechnicalIndicators.calculateSMA(previousCloses, params.slowMA)
        val spreadPercent = if (slowMA > 0.0) (fastMA - slowMA) / slowMA * 100.0 else 0.0
        val spreadEnough = spreadPercent >= minSpreadPercent()
        val trendOk = !strategyConfig.doubleMAConfig.requireSlowMAUp ||
                (prevSlowMA > 0.0 && slowMA >= prevSlowMA && price >= slowMA)
        val costOk = minRewardCostMultiple() <= 0.0 ||
                params.takeProfitPercent >= estimatedRoundTripCostPercent(price) * minRewardCostMultiple()
        val buySetup = fastMA > slowMA &&
                spreadEnough &&
                trendOk &&
                costOk &&
                rsi < 70 &&
                !isLateBuyBlocked(bar.timestamp) &&
                cooldownBarsRemaining <= 0
        val sellSetup = fastMA < slowMA

        if (!hasPosition) {
            sellConfirmBars = 0
            buyConfirmBars = if (buySetup) buyConfirmBars + 1 else 0
            if (buyConfirmBars >= confirmationBars()) {
                hasPosition = true
                positionEntryPrice = price
                entryTime = Date()
                buyConfirmBars = 0
                positionStopLossPrice = positionEntryPrice * (1 - params.stopLossPercent / 100.0)
                positionTakeProfitPrice = positionEntryPrice * (1 + params.takeProfitPercent / 100.0)
                emitSignal(
                    SignalType.BUY, price, params.lotSize,
                    "${barPeriodMinutes()}分钟K线多头确认: 快MA(${params.fastMA})高于慢MA(${params.slowMA})，差值${String.format("%.2f", spreadPercent)}%"
                )
            }
            return
        }

        buyConfirmBars = 0

        if (price <= positionStopLossPrice) {
            hasPosition = false
            enterCooldown()
            emitSignal(
                SignalType.STOP_LOSS, price, params.lotSize,
                "止损触发: 亏损${params.stopLossPercent}%"
            )
            return
        }

        if (price >= positionTakeProfitPrice) {
            hasPosition = false
            enterCooldown()
            emitSignal(
                SignalType.TAKE_PROFIT, price, params.lotSize,
                "止盈触发: 盈利${params.takeProfitPercent}%"
            )
            return
        }

        sellConfirmBars = if (sellSetup) sellConfirmBars + 1 else 0
        if (sellConfirmBars >= confirmationBars()) {
            hasPosition = false
            enterCooldown()
            emitSignal(
                SignalType.SELL, price, params.lotSize,
                "${barPeriodMinutes()}分钟K线空头确认: 快MA(${params.fastMA})低于慢MA(${params.slowMA})"
            )
        }
    }

    private fun updateStopLossTakeProfit(price: Double) {
        val currentProfit = (price - positionEntryPrice) / positionEntryPrice * 100.0
        if (currentProfit >= 3.0) {
            val newSL = positionEntryPrice * (1 + 0.01)
            if (newSL > positionStopLossPrice) {
                positionStopLossPrice = newSL
            }
        }
        if (currentProfit >= 5.0) {
            val newTP = positionEntryPrice * (1 + 0.10)
            if (newTP > positionTakeProfitPrice) {
                positionTakeProfitPrice = newTP
            }
        }
    }
}
